### ISSUE TITLE
`TMB::normalize` issues

### ISSUE THREAD

**Issue by James-Thorson** [kaskr/adcomp#264]
Kasper,

Thanks again for attending the CAPAM fishery-dependent standardization workshop last week, and for explaining how to use the un-normalized GMRF features using code on the CPP side:

``` tmb
int normalize = 0;  
jnll += GMRF( Q, bool(normalize) );
```

and then on the R side

```R
obj_normalized = TMB::normalize(obj, flag="include_data", value=FALSE)
```

As you showed last week, a simple log-Gaussian-Cox model appears to run ~10 fold faster and with 1/10th as much memory using this compared with the vanilla GMRF implementation.

However, I have two responses to the current implementation:

1.  It appears that the compiled object `obj_normalized` has a different location for several tagged inputs than the vanilla object `obj`.  For example, `obj$report()` appears to be moved to `obj_normalized$env$report()`.  Moving their location within the compiled object obviously breaks the standard API for TMB.

2.  More importantly, I have confirmed that small applications of package `VAST` are slower doing the normalization in R rather than TMB, sometimes by 3-4 fold.  It *appears* that this is caused because the `obj_normalized` does not correctly save the `last_par_best` for the inner optimization step, so that in VAST it keeps doing 20-30 inner-optimization steps when calling `obj_normalized$fn`, rather than 20-30 inner steps the first time `obj$fn` is called and then 2-3 inner steps in subsequent calls of `obj$fn` using vanilla implementation of GMRF.  For larger `VAST` models, similarly, doing the normalization in R has essentially no improvement in speed, because the increase in inner-optimization steps appears to be offset by the improved speed from doing normalization in R.

Anyway, the normalization-in-R approach seems very promising, so I welcome any feedback on how to help test or improve the implementation!

**Comment by kaskr** [comment]
@James-Thorson 
Thanks for testing it Jim,

1. It seems you're right. The normalized object lacks `report` and also `simulate`.

2. I think the `last.par.best` **is** being correctly set. At least I couldn't reproduce this problem. Could you perhaps post the output you get from the different objects ? It's definitely a bug if the normalized object uses 20-30 steps for each evaluation of `fn`.

**Comment by James-Thorson** [comment]
Kasper,

I checked again this morning, and I think that the number of inner-optimization steps in `VAST` varies greatly depending upon where it is in the outer-optimizer.  So I can't reproduce an easy example of #2 above, and probably shouldn't have made the claim without better evidence!  

but if you were willing to fix #1 above, that would be helpful, and I will continue to work with NOAA folks to try and explore why the normalized-in-R version is so much slower on simple VAST models.

**Comment by kaskr** [comment]
@James-Thorson 
Missing slots have been added to normalized object.
